src/Microsoft.SqlTools.ServiceLayer/TableDesigner/TableDesignerValidator.cs

```diff
@@ -0,0 +1,263 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System.Collections.Generic;
+using Microsoft.Data.Tools.Sql.DesignServices.TableDesigner;
+using ValidationError = Microsoft.SqlTools.ServiceLayer.TableDesigner.Contracts.TableDesignerValidationError;
+namespace Microsoft.SqlTools.ServiceLayer.TableDesigner
+{
+    public static class TableDesignerValidator
+    {
+        private static List<ITableDesignerValidationRule> Rules = new List<ITableDesignerValidationRule>();
+
+        static TableDesignerValidator()
+        {
+            Rules.Add(new IndexMustHaveColumnsRule());
+            Rules.Add(new ForeignKeyMustHaveColumnsRule());
+            Rules.Add(new ColumnCanOnlyAppearOnceInForeignKeyRule());
+            Rules.Add(new ColumnCanOnlyAppearOnceInIndexRule());
+            Rules.Add(new NoDuplicateColumnNameRule());
+            Rules.Add(new NoDuplicateConstraintNameRule());
+            Rules.Add(new NoDuplicateIndexNameRule());
+        }
+
+        /// <summary>
+        /// Validate the table and return the validation errors.
+        /// </summary>
+        public static List<ValidationError> Validate(TableViewModel table)
+        {
+            var errors = new List<ValidationError>();
+            foreach (var rule in Rules)
+            {
+                errors.AddRange(rule.Run(table));
+            }
+            return errors;
+        }
+    }
+
+    public interface ITableDesignerValidationRule
+    {
+        List<ValidationError> Run(TableViewModel table);
+    }
+
+    public class IndexMustHaveColumnsRule : ITableDesignerValidationRule
+    {
+        public List<ValidationError> Run(TableViewModel table)
+        {
+            var errors = new List<ValidationError>();
+            for (int i = 0; i < table.Indexes.Items.Count; i++)
+            {
+                var index = table.Indexes.Items[i];
+                if (index.Columns.Count == 0)
+                {
+                    errors.Add(new ValidationError()
+                    {
+                        Message = string.Format("Index '{0}' does not have any columns associated with it.", index.Name),
+                        PropertyPath = new object[] { TablePropertyNames.Indexes, i }
```

Comment:
If the path is just strings and ints any reason you couldn't just make it `string[]` instead?

Reply:
nothing major, I think I was trying to avoid the type casting. in ADS it is more natural since typescript supports union types.